pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -212,36 +218,56 @@ public void reloadSegment(String tableNameWithType, String segmentName, boolean
   }
 
   @Override
-  public void reloadAllSegments(String tableNameWithType, boolean forceDownload) {
+  public void reloadAllSegments(String tableNameWithType, boolean forceDownload,
+      @Nullable Semaphore refreshThreadSemaphore)
+      throws Exception {
     LOGGER.info("Reloading all segments in table: {}", tableNameWithType);
     TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
     Preconditions.checkNotNull(tableConfig);
-
     Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, tableNameWithType);
-
     List<String> failedSegments = new ArrayList<>();
-    Exception sampleException = null;
     List<SegmentMetadata> segmentsMetadata = getAllSegmentsMetadata(tableNameWithType);
-    for (SegmentMetadata segmentMetadata : segmentsMetadata) {
+    ExecutorService workers = Executors.newCachedThreadPool();
+    final AtomicReference<Exception> sampleException = new AtomicReference<>();
+    //calling thread hasn't acquired any permit so we don't reload any segments using it.
+    CompletableFuture.allOf(segmentsMetadata.stream().map(segmentMetadata -> CompletableFuture.runAsync(() -> {
       try {
+        acquireSema("ALL", refreshThreadSemaphore);
         reloadSegment(tableNameWithType, segmentMetadata, tableConfig, schema, forceDownload);
       } catch (Exception e) {
         String segmentName = segmentMetadata.getName();
         LOGGER.error("Caught exception while reloading segment: {} in table: {}", segmentName, tableNameWithType, e);
         failedSegments.add(segmentName);
-        sampleException = e;
+        sampleException.set(e);
+      } finally {
+        if (refreshThreadSemaphore != null) {
+          refreshThreadSemaphore.release();
+        }
       }
-    }
+    }, workers)).toArray(CompletableFuture[]::new)).get();
 
-    if (sampleException != null) {
+    if (sampleException.get() != null) {
       throw new RuntimeException(
           String.format("Failed to reload %d/%d segments: %s in table: %s", failedSegments.size(),
-              segmentsMetadata.size(), failedSegments, tableNameWithType), sampleException);
+              segmentsMetadata.size(), failedSegments, tableNameWithType), sampleException.get());
     }
-
     LOGGER.info("Reloaded all segments in table: {}", tableNameWithType);
   }
 
+  private void acquireSema(String context, Semaphore refreshThreadSemaphore)
```

Comment:
The first argument is `segmentName`

(optional) This part is duplicated. One way to avoid duplicate code is to add a `SegmentRefreshSemaphore` class to wrap this logic.

Reply:
Created a wrapper class.